bob-access/src/extractor.rs

```diff
@@ -0,0 +1,117 @@
+use std::net::SocketAddr;
+
+use crate::{credentials::Credentials, error::Error};
+use axum::extract::RequestParts;
+use http::{Extensions, HeaderMap, Request};
+use tonic::{transport::server::TcpConnectInfo, Request as TonicRequest};
+
+pub trait Extractor {
+    fn headers(&self) -> Option<&HeaderMap>;
+    fn extensions(&self) -> Option<&Extensions>;
+}
+
+pub trait ExtractorExt {
+    fn extract(&self) -> Result<Credentials, Error>;
+    fn extract_basic(
+        &self,
+        _header_map: &HeaderMap,
+        _addr: Option<SocketAddr>,
+    ) -> Result<Option<Credentials>, Error> {
+        Ok(None)
+    }
+    fn extract_token(
+        &self,
+        _header_map: &HeaderMap,
+        _addr: Option<SocketAddr>,
+    ) -> Result<Option<Credentials>, Error> {
+        Ok(None)
+    }
+}
+
+impl<T: Extractor> ExtractorExt for T {
+    fn extract(&self) -> Result<Credentials, Error> {
+        let header_map = if let Some(header_map) = self.headers() {
+            header_map
+        } else {
+            return Ok(Credentials::default());
+        };
+        let addr = self
+            .extensions()
+            .and_then(|ext| ext.get::<TcpConnectInfo>()?.remote_addr());
+        let basic_credentials = self.extract_basic(header_map, addr)?;
+        let token_credentials = self.extract_token(header_map, addr)?;
+        match (basic_credentials, token_credentials) {
+            (Some(_), Some(_)) => Err(Error::MultipleCredentialsTypes),
+            (Some(basic_credentials), None) => Ok(basic_credentials),
+            (None, Some(token_credentials)) => Ok(token_credentials),
+            _ => Ok(Credentials::builder()
+                // .with_address(req.remote_addr())
```

Comment:
Why not leave `with_address` here and remove from specific extraction functions?

Reply:
`Builder` can be passed into `extract_*` functions, so this one will contain common setup for builder and specific functions will add either username+password or token

Reply:
This branch is for `(None, None)` case. As I understand, it is triggered with inter-node communication and also when no authorization is enabled. For the first case `remote_addr` is required. For the second case we should not try to extract any credentials at all